lib/spack/spack/version/version_types.py

```diff
@@ -61,27 +63,30 @@ def __str__(self):
     def __repr__(self) -> str:
         return f'VersionStrComponent("{self}")'
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
         return isinstance(other, VersionStrComponent) and self.data == other.data
 
-    def __lt__(self, other):
-        lhs_inf = isinstance(self.data, int)
+    # ignore typing for certain parts of these methods b/c a) they are performance-critical, and
+    # b) mypy isn't smart enough to figure out that if l_inf and r_inf are the same, comparing
+    # self.data and other.data is type safe.
+    def __lt__(self, other: object) -> bool:
+        l_inf = isinstance(self.data, int)
         if isinstance(other, int):
-            return not lhs_inf
-        rhs_inf = isinstance(other.data, int)
-        return (not lhs_inf and rhs_inf) if lhs_inf ^ rhs_inf else self.data < other.data
-
-    def __le__(self, other):
-        return self < other or self == other
+            return not l_inf
+        r_inf = isinstance(other.data, int)  # type: ignore
+        return (not l_inf and r_inf) if l_inf ^ r_inf else self.data < other.data  # type: ignore
 
-    def __gt__(self, other):
-        lhs_inf = isinstance(self.data, int)
+    def __gt__(self, other) -> bool:
```

Comment:
```suggestion
    def __gt__(self, other: object) -> bool:
```

Reply:
Done in [86e2dc1](https://github.com/spack/spack/pull/47781/commits/86e2dc15203e7439a48edcae5ae5c64cbd900d20)